Integrate Firebase Authentication

calebschoepp opened the pull request:
## Summary
After a lot more time figuring out 2-way binding than I care to admit, I've added basic authentication to our app! There's two key parts to the changes. First, the auth activity. Second, the authenticated activity.

The `AuthActivity` is pretty straightforward. It really is just a login screen and a register screen that both hook into Firebase Auth. I wired it up to bring you to a logout button just to test the functionality.

The `AuthenticatedActivity` is the cool bit. Every activity (except `AuthActivity` b/c of infinite recursion reasons :tongue: ) extends it. In short what it does is redirect to the login screen if the user is not logged in. This means that any possible entry point to our app will be secure (launcher, notification, etc.). Please ask any questions you have about this architecture.

## Changes
- Rename a bunch of things. See below.
- Implement login, register, and an example logout
- Implement an abstract activity to lock everything behind authentication (basically a middle-ware)
- Remove actionbar
- Update to use Java 8 to let us use lambdas
- Added a `SingleLiveEvent` class. PLEASE QUESTIONS ABOUT THIS, b/c it is a super valuable pattern.
- The UI is a work in progress

## Open Questions
- Is everyone okay with the renaming I did?
    - Sign In -> Login (Feels simpler to me)
    - Sign Up -> Create Account / Register (Feels simpler to me)
    - `LoginActivity` -> `AuthActivity` (Feels more accurate to me)
- People okay with me bumping to Java 8 for lambdas?
- I didn't collect first name and last name b/c we are short on screen space and I don't think it is actually in the requirements. This okay?

## TODO
- Build out the logic to connect a `FirebaseUser` to our custom Firestore `User` entity.
- Test `AuthViewModel`
- A number of TODOs littered throughout my changes

[Video of UI](https://photos.google.com/share/AF1QipPnCJfdp5dpNGaAbKAiTGsYpPx3I-p6HV2KGxxXFyRF73cKl9nTjW8rkiHlYaJvUw?key=N2lZQklFd3NmZTRELVo3cnpaMElDRnpmY2lnbWd3)

cyrusdiego commented on app/src/main/java/com/example/unlibrary/auth/AuthViewModel.java:
```diff
@@ -0,0 +1,144 @@
+
+    public MutableLiveData<String> getEmail() {
+        if (mEmail == null) {
+            mEmail = new MutableLiveData<String>();
```
Nit: Can replace `<String>` with `<>` (From Android IDE: Explicit type argument String can be replaced with <> ), same for all getters below

cyrusdiego commented on app/src/main/java/com/example/unlibrary/auth/AuthViewModel.java:
```diff
@@ -0,0 +1,144 @@
+
+public class AuthViewModel extends ViewModel {
+
+    private MutableLiveData<String> mEmail = new MutableLiveData<>();
```
Is it necessary to initialize these MutableLiveData variables here if you do so in the getters?

calebschoepp replied:
Good question. You're right that they get initialized in the getters. However, I still initialize them here because it is possible that the user clicks a button before the get function is called and it would throw a null exception error when trying to validate the data. Pre-initializing them here avoids that.

cyrusdiego replied:
Ah makes sense, thanks for clarifying.

calebschoepp replied:
No problem. This is one of the things I was struggling with for so long :rofl:

cyrusdiego commented on app/src/main/java/com/example/unlibrary/auth/AuthViewModel.java:
```diff
@@ -0,0 +1,144 @@
+
+    public void login() {
+        // Validate data
+        if (mEmail.getValue() == null || mEmail.getValue().isEmpty()) {
```
Would probably want to set up thorough validation of email (missing '@', invalid email address, etc.) . Maybe it's something that can be done in Firebase Auth side rather than doing say Regex here.

cyrusdiego replied:
Also, Material's EditText has built in "error" messages. I think those would be better at showing an invalid value passed instead of Toasts. [Link](https://material.io/develop/android/components/text-fields) (Skip to "Adding errors to a text field")

calebschoepp replied:
Agreed that more validation is needed. Couple things:
- I tested registering with an invalid email and firebase stops it, but it would be nice to detect it client side and let the user know in advance
- Validation also needs to be added to things like the username (globally unique, no spaces, no crazy unicode crap, etc.)
- IMO this is outside the scope of this PR. Since this PR is already big is it okay if I file a ticket and include this validation in a follow-up PR?

calebschoepp replied:
Re: Error messages

I agree that text field errors is a better idea. I'll work on that. You fine if I still show `Invalid username or password.`  in a toast since it applies to both fields?

cyrusdiego replied:
Re: validation
I agree it should be in a separate PR, now that you point out more details and now that I think about it, it'll be a decent amount of work to add logic on the client side. Yeah if you can just file another ticket for it, thanks!

Re: Error Messages
Yeah that should be fine to show those in a toast.

calebschoepp replied:
Re: Error Messages

Now that I'm looking at the implementation for this too, I'm wondering if I could push it out of this PR and into my follow-up PR where I add validation. The errors are tightly coupled to the validation and it would also be nice if the errors were live with the text input. You alright if I punt this to the ticket I filed and add it in a follow-up PR?

cyrusdiego replied:
Yeah that's fine by me :)

cyrusdiego commented on app/src/main/res/layout/fragment_login.xml:
```diff
@@ -0,0 +1,109 @@
+                    android:layout_width="match_parent"
+                    android:layout_height="wrap_content"
+                    android:layout_margin="@dimen/text_margin"
+                    android:hint="Email">
```
Move hardcoded strings to the string resource file

cyrusdiego commented on app/src/main/res/layout/fragment_login.xml:
```diff
@@ -0,0 +1,109 @@
+                    android:layout_width="match_parent"
+                    android:layout_height="wrap_content"
+                    android:layout_margin="@dimen/text_margin"
+                    android:hint="Password"
```
Move hardcoded strings to the string resource file

cyrusdiego commented on .github/workflows/main.yml:
```diff
@@ -53,3 +53,4 @@ jobs:
           arg-spec: '.github/test-lab-config.yml:android-pixel-4'
         env:
           SERVICE_ACCOUNT: ${{ secrets.SERVICE_ACCOUNT }}
+        continue-on-error: true
```
Should this change be in this PR? I thought this was already merged into master with your test PR

calebschoepp replied:
This is only here because I hadn't rebased to master yet. This is now resolved.

cyrusdiego commented on app/src/main/java/com/example/unlibrary/auth/AuthViewModel.java:
```diff
@@ -0,0 +1,144 @@
+        mFragmentNavigationEvent.setValue(new LoginFragment());
+    }
+
+    public void register() {
```
Look at comments for the login() function, applies here as well

calebschoepp replied:
Resolving because we've addressed these concerns on `login()` thread.

cyrusdiego commented on app/src/main/java/com/example/unlibrary/util/SingleLiveEvent.java:
```diff
@@ -0,0 +1,62 @@
+package com.example.unlibrary.util;
+
+// https://github.com/NordicSemiconductor/Android-nRF-Blinky/blob/master/app/src/main/java/no/nordicsemi/android/blinky/viewmodels/SingleLiveEvent.java
```
In terms of citing, is this enough? Do we have to put in like licensing, date, etc? Not sure how nitpicky they'll be but I think I remember the prof talking about citing code in Assignment 1

calebschoepp replied:
Yeah I'm not sure. This sounds like a bigger conversation we need to have as a team about how we want do citations. For now, this link will be enough to get us back to the info we might need. I'll add a TODO as a reminder.

cyrusdiego commented on app/src/main/res/layout/activity_profile.xml:
```diff
@@ -6,4 +6,14 @@
+        android:layout_width="wrap_content"
+        android:layout_height="wrap_content"
+        android:onClick="logout"
+        android:text="Logout"
```
This is just a placeholder for now so up to you but could move the hardcoded string to the string resource file if you want.

calebschoepp replied:
Just going to leave it hardcoded. Don't want us to forget about the string resource if we no longer need it.

cyrusdiego reviewed (changes requested):
Added some comments throughout, overall it looks good to me :D Aside from the general comments should also have documentation (Javadoc style) and unit tests (if/where applicable)

cyrusdiego commented:
>Is everyone okay with the renaming I did?
Sign In -> Login (Feels simpler to me)
Sign Up -> Create Account / Register (Feels simpler to me)
LoginActivity -> AuthActivity (Feels more accurate to me)

I think the renames are good, a lot clearer tbh

> People okay with me bumping to Java 8 for lambdas?

I'm good with it, lol was going to bump it to Java 8 myself to use lambdas 

> I didn't collect first name and last name b/c we are short on screen space and I don't think it is actually in the requirements. This okay?

Should be fine, would be a fairly simple fix if the spec's change to require first and last name

cyrusdiego reviewed (commented):
Recent changes looks good! 
Can probably add the Javadoc comments in this PR but not sure if we'd want to have a separate PR to actually generate the HTML Javadoc? Up to you.

calebschoepp commented:
> Can probably add the Javadoc comments in this PR but not sure if we'd want to have a separate PR to actually generate the HTML Javadoc? Up to you.

Yep working on adding the comments rn. Probably won't bother with generating the docs. We can do that closer to the deadline or perhaps set it up as a step in our CI/CD.

calebschoepp commented:
@cyrusdiego I've finished adding docs and unit tests. Just looking for the LGTM :)

cyrusdiego commented on app/src/main/java/com/example/unlibrary/EntryActivity.java:
```diff
@@ -0,0 +1,21 @@
+import com.example.unlibrary.profile.ProfileActivity;
+import com.example.unlibrary.util.AuthenticatedActivity;
+
+public class EntryActivity extends AuthenticatedActivity {
```
Missing class documentation

cyrusdiego commented on app/src/main/java/com/example/unlibrary/EntryActivity.java:
```diff
@@ -0,0 +1,21 @@
+public class EntryActivity extends AuthenticatedActivity {
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
```
missing method documentation

cyrusdiego commented on app/src/main/java/com/example/unlibrary/EntryActivity.java:
```diff
@@ -0,0 +1,21 @@
+package com.example.unlibrary;
```
Missing source file header

cyrusdiego commented on app/src/main/java/com/example/unlibrary/auth/AuthActivity.java:
```diff
@@ -0,0 +1,63 @@
+package com.example.unlibrary.auth;
```
Missing source file header

cyrusdiego commented on app/src/main/java/com/example/unlibrary/auth/AuthViewModel.java:
```diff
@@ -0,0 +1,183 @@
+package com.example.unlibrary.auth;
```
Missing source file header

cyrusdiego commented on app/src/main/java/com/example/unlibrary/auth/LoginFragment.java:
```diff
@@ -0,0 +1,33 @@
+package com.example.unlibrary.auth;
```
Missing source file header

cyrusdiego commented on app/src/main/java/com/example/unlibrary/auth/RegisterFragment.java:
```diff
@@ -0,0 +1,36 @@
+package com.example.unlibrary.auth;
```
Missing source file header

cyrusdiego commented on app/src/main/java/com/example/unlibrary/util/AuthenticatedActivity.java:
```diff
@@ -0,0 +1,48 @@
+package com.example.unlibrary.util;
```
Missing source file header

cyrusdiego commented on app/src/main/java/com/example/unlibrary/util/SingleLiveEvent.java:
```diff
@@ -0,0 +1,63 @@
+package com.example.unlibrary.util;
```
Missing source file header, can probably put citation in here too

cyrusdiego commented on app/src/main/java/com/example/unlibrary/util/SingleLiveEvent.java:
```diff
@@ -0,0 +1,63 @@
+
+    @MainThread
+    @Override
+    public void observe(@NonNull final LifecycleOwner owner, @NonNull final Observer<? super T> observer) {
```
missing method documentation

cyrusdiego commented on app/src/main/java/com/example/unlibrary/util/SingleLiveEvent.java:
```diff
@@ -0,0 +1,63 @@
+
+    @MainThread
+    @Override
+    public void setValue(@Nullable final T t) {
```
missing method documentation

cyrusdiego commented on app/src/test/java/com/example/unlibrary/auth/AuthViewModelTest.java:
```diff
@@ -0,0 +1,94 @@
+package com.example.unlibrary.auth;
```
Missing source file header

cyrusdiego commented on app/src/test/java/com/example/unlibrary/auth/AuthViewModelTest.java:
```diff
@@ -0,0 +1,94 @@
+
+import static org.junit.Assert.assertEquals;
+
+public class AuthViewModelTest {
```
I think for documenting the tests we don't need method comments but maybe put in a class comment? I'll leave this one up to you though, I didn't document my test for the recent lab so I'm not sure if we absolutely need to document unit test

calebschoepp replied:
Yeah, probably adequate just to add a class level comment describing what it tests.

cyrusdiego commented on app/src/test/java/com/example/unlibrary/helper/LiveDataTestUtil.java:
```diff
@@ -0,0 +1,43 @@
+package com.example.unlibrary.helper;
```
Missing source file header, can put citation here

calebschoepp replied:
Will leave citation where it is at rn b/c we need to decide on a their formatting.

cyrusdiego reviewed (changes requested):
Regarding documentation, I'm looking at the Javadoc lab slides ([link](https://eclass.srv.ualberta.ca/pluginfile.php/6121307/mod_resource/content/0/Code%20Conventions%20%20Javadoc%20%28Schoepp%29.pdf)) that were posted for the standards we'd have to follow. I think for the most part, all that's missing are the "Beginning Comments" at the top of each source file: 
![image](https://user-images.githubusercontent.com/32345990/96351581-9150a200-1079-11eb-9c6a-74eb96a4d7e9.png)

Sorry for being nitpicky with this D: I just read through the slides from the lab and I'm worried they're gonna be nitpicky with documenting and such.

calebschoepp commented:
Hey no, I like a thorough review. I originally left them out because I thought the TA said we didn't need them, but I think you're actually right. I'll go back and add them after today, I'm currently busy.

calebschoepp commented:
@cyrusdiego I have addressed all of your comments. Once again, don't worry about all the nits. Hopefully, this PR can serve as a goal post for all of the standards we want to enforce in our code.

I left the copyright/licenses as TODOs b/c IMO the team needs to have a discussion about formatting for this and I'd like to merge this behemoth PR sooner than later to minimize merge conflict headaches for other teamates.

cyrusdiego reviewed (approved):
LGTM! Thanks for working on this and doing all the documentation and unit tests :D

armiantos commented:
Closes #32
